src/messenger/messenger.cpp

```diff
@@ -418,25 +389,29 @@ void Messenger::RunTask() {
 }
 
 void Messenger::Terminate() {
+  if (is_messenger_running_) {
+    // Shut down the ZeroMQ context. This causes all existing sockets to abort with ETERM.
+    zmq_ctx_->shutdown();
+    MessengerUtil::DeleteLocalIPC(port_);
+    MESSENGER_LOG_INFO(fmt::format("[PID={}] Messenger terminated.", ::getpid()));
+  }
   is_messenger_running_ = false;
-  // Shut down the ZeroMQ context. This causes all existing sockets to abort with ETERM.
-  zmq_ctx_->shutdown();
-  MESSENGER_LOG_INFO(fmt::format("[PID={}] Messenger terminated.", ::getpid()));
 }
 
 void Messenger::ListenForConnection(const ConnectionDestination &target, const std::string &identity,
                                     CallbackFn callback) {
-  std::lock_guard lock(routers_add_mutex_);
+  std::unique_lock lock(routers_add_mutex_);
```

Comment:
I'm a little confused by this part. If we're taking the time to get a latch, why not just have the latch be for `routers_` and insert it directly into that? Is there something I'm missing here?